### DISCUSSION TITLE
Question on adding libraries and confusion on Make, makelists and so on

### DISCUSSION THREAD

**Discussion by Mair** [Mair/esp32-course#113] [Course Qeustions]
The following was asked by a student...

Hello Mr Mair,
I hope this email finds you well.

Your esp32 course is really wonderful,
But the esp idf project structure still gives me headaches.
Adding files, folders, libraries"lvgl".
Mainly, because I have no clue with makelist and cmake.
And till now I can only work with esp by modifying the examples.

May you consider adding a video or two on this topic -cmake makelist sdkconfig- to the course?
I don't know what makes the idf powershell recognize a folder as an esp idf workspace or not.
And may you refer me to the required resources to study?

I can't wait to hear from you back.

Many thanks,
Studend

**Comment by Mair** [comment]
I agree it can be confusing, the good news is that once you understand the basics you don’t need to worry about it too much. Most of it is taken care of for you when using the idf.py commands
Here are a few things that will hopefully sort it out for you.

1.	The IDF used to use “Make”. for “Make” to work you needed to have a Makefile in the root and a component.mk in the main folder. 
2.	“Make” is no longer used so these files can be deleted / ignored
3.	“Make” was replaced with “CMake”’
4.	“CMake” uses a CMakelists.txt in the root of the project and in the main folder

![image](https://user-images.githubusercontent.com/2477839/169794756-fef5b9da-2260-4748-87af-381511e8f6ce.png)

 

5.	Inside the main folder, the CMakeLists.txt uses idf_component_register to tell the compiler which files to compile using SRC (see picture on the left) and where the header files are using INCLUDE_DIR. SRC expects all the c file in your project “ main.c” “file1.c” “file2.c” etc. INCLUDE_DIR expects a directory where the header files are “.” (“.” means local directory, in this case the “main” directory)
6.	If you create a components directory in the root of the project and then put a subfolder in there such as LVGL. There will need to be a CMakeLists.txt file in there that does the same thing. It tells the complier where the source files and header files are
7.	The root CMakeLists.txt. has a line that says include($ENV{IDF_PATH}/tools/cmake/project.cmake)
8.	This is pointing to a massive file that tells the compiler where all the IDF components are. 
![image](https://user-images.githubusercontent.com/2477839/169794782-099e4fef-036e-4885-b4cb-85c0b8043fc9.png)

 
9.	Interestingly each one of these components also has a CMakeLists.txt file that tells the compiler where the c files and h files are just like our main/CMakeLits.txt in our project does.
10.	So, to summarize a component library is a directory that has a CMakeLists.txt file in it that tells the compiler what the c and h files are.
11.	In out root CMakeLists.txt we can also add other component libraries that are not included in the IDF or in our project. As an example you can see https://learnesp32.com/videos/12/1_example%20connection
12. In the course I use a slightly older way to add files it still works and the exact same principle applies see https://learnesp32.com/videos/3/3_referencing%20c%20and%20h%20files and https://learnesp32.com/videos/3/4_idf%20components%20-%20other%20source%20files
The official documentation can be found at https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system.html

**Comment by ashrafkamel5** [comment]
Many thanks for your answer!
Surely, it made things much clear.
